### PR TITLE
add support for running agent helm chart on persistent volume

### PR DESCRIPTION
**Pull request by knisbet**
Problem: In teleport cloud we are setup to automatically rotate the join tokens to the teleport cluster. The helm chart isn't currently compatible with this as the join token gets added as a kubernetes secret but is not kept in sync with the cluster.

This is an addition to optionally allow the pods to be deployed as a statefulset with a persistent volume from kubernetes/cloud to maintain the state directory through pod restarts. I believe the Teleport team is also looking at using the kubernetes API to store this state, but I don't know if/when that will be available, so this seemed like a reasonable solution.

Submitting it here if the core team is interested, otherwise we'll have to fork the chart for use in the cloud. 

I originally approached this as a bunch of replacements on the deployment file, changing the API and other parameters inline, but the approach seemed fragile with the number of tweaks. So I went with this approach to disable the deployment and add the statefulset if persistent storage is enabled, which largely duplicates the object definition that will require certain updates to be duplicated in both places.

Also, not super familiar with the linting approach used with helm, so I think I added the relevant tests.

### CODE REVIEW

**Comment by webvictim** [comment]
https://github.com/gravitational/teleport/pull/7096 does a similar thing but seems a lot simpler (although doesn't add linting as your PR does etc) - what's the incentive for using a `StatefulSet` instead?

**Comment by knisbet** [comment]
It's not totally clear to me how #7096 is meant to work, if it's attempting to use a localDir for persistence, or some other way to map a single volume. Since the deployment controller doesn't normally account for persistence, I'm not really sure if this is a recommended model, and won't run into problems when say bumping the container tag where the deployment controller would then want to replace the pod with a new one by temporarily using two instances using the same persistence store.

The StatefulSet controller is more geared towards handling persistence, so in this approach the prereq is for the cluster to be connected to a cloud environment / storage controller with a storage class defined, and the volumes will be provisioned by kubernetes in the cloud environment by installing the chart. The statefulset controller also understands and won't try and run more than one pod connected to the same volume, or schedule pods in other AZs than the volume. Will probably also be more appropriate if we want to try and create redundant deployments through increasing the replicas. 

the TLDR, StatefulSet is geared towards applications with state, Deployments are geared towards stateless (or more specifically ephemeral pods).

**Comment by webvictim** [comment]
@klizhentas @awly What are your thoughts on this from a product standpoint? Are we looking to prioritise the work to use the Kubernetes API for certificate storage, or should we look to merge a chart-based solution for this in the short term?

**Comment by awly** [comment]
Storing state in k8s API (https://github.com/gravitational/teleport/issues/4832) is not on the horizon any time soon.
Adding support for persistent volumes in our chart in the meantime makes sense.
The only reason I didn't add it originally was cross-platform compatibility.

@knisbet will you change work across at least AWS/GCP/Azure?

**Comment by knisbet** [comment]
> @knisbet will you change work across at least AWS/GCP/Azure?

It should...

This really just leverages the kubernetes support for persistent volumes, which can integrate with various cloud providers and different plugins. So it just requires having or configuring a storageclass (https://kubernetes.io/docs/concepts/storage/storage-classes/) that matches the particular cloud provider and when installing the helm chart point at the storageclass created by the customer that will then have cloud provider specific parameters.

File: examples/chart/teleport-kube-agent/values.yaml [hunk 1]
```diff
@@ -57,6 +57,17 @@ podSecurityPolicy:
...
+################################################################
+# Values that must be provided if using persistent storage for Teleport state
+# enabled: Whether to use persistent storage
+# storageClassName: The name of the kubernetes storage class to use when creating volumes. See https://kubernetes.io/docs/concepts/storage/storage-classes/
+# requests: The size of the volume to request from the persistent storage system
+################################################################
+storage:
+  enabled: false
+  storageClassName: ""
+  requests: 128Mi
```
Comment (awly) [inline examples/chart/teleport-kube-agent/values.yaml:69]: I actually have no idea how big the teleport storage needs are.
Do you have any numbers on the real-world size?
If this needs to be increased after deployment, will existing data be preserved?

File: examples/chart/teleport-kube-agent/templates/deployment.yaml [hunk 2]
```diff
@@ -65,6 +66,10 @@ spec:
     {{- toYaml .Values.extraVolumeMounts | nindent 8 }}
   {{- end }}
 {{- end }}
+      {{- if .Values.nodeSelector }}
+      nodeSelector:
+        {{- toYaml .Values.nodeSelector | nindent 6 }}
```
Comment (awly) [inline examples/chart/teleport-kube-agent/templates/deployment.yaml:71]:
```suggestion
        {{- toYaml .Values.nodeSelector | nindent 8 }}
```

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml [hunk 3]
```diff
@@ -0,0 +1,140 @@
+{{- if .Values.storage.enabled }}
+{{- if .Values.teleportVersionOverride }}
+  {{- $_ := set . "teleportVersion" .Values.teleportVersionOverride }}
+{{- else }}
+  {{- $_ := set . "teleportVersion" .Chart.AppVersion }}
```
Comment (awly) [inline examples/chart/teleport-kube-agent/templates/statefulset.yaml:5]:
```suggestion
  {{- $_ := set . "teleportVersion" .Chart.Version }}
```

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,140 @@
+{{- if .Values.storage.enabled }}
```
Comment (awly) [inline examples/chart/teleport-kube-agent/templates/statefulset.yaml:1]: it's unfortunate that 90% of this file is identical to `deployment.yaml`
we will have drift between the contents of these files over time.

let's at least add a comment at the top of both files as a reminder to update the other one.

**Review by awly — COMMENTED** [review]
Overall I like this change.
To release it as a part of our public Helm chart we should at least document how to set this up (maybe link to cloud provider-specific docs if available).
Also, if it's possible to migrate from non-stateful setup, we should document how. If it's not possible, we need a warning somewhere.

File: examples/chart/teleport-kube-agent/values.yaml (hunk 1 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/values.yaml:69]: > I actually have no idea how big the teleport storage needs are.
> Do you have any numbers on the real-world size?
Last 7 days I've had this deployed / idle it's used 19MiB. I may have assumed incorrectly that because this is a node agent and logs/audit/etc are meant to be centralized that the default for this volume can be relatively small.

> If this needs to be increased after deployment, will existing data be preserved?
It depends on how the storage class is setup and presumably whether the block driver supports it. The storage class has a boolean field called allowVolumeExpansion, and if it's set, the volume can be expanded without data loss. It will interrupt the pod temporarily since the volume needs to be unmapped and the filesystem expanded.

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml (hunk 4 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/templates/statefulset.yaml:1]: Yea, TBH I originally wrote it as an ifdef's in the one file for just the changes, but that also seemed like a logical mess to support the 10% differences. IDK which way is better, both seem likely to produce issues.

**Comment by knisbet** [comment]
> Overall I like this change.
> To release it as a part of our public Helm chart we should at least document how to set this up (maybe link to cloud provider-specific docs if available).

Are these agent helm charts documented somewhere in general? I see a few references to the other chart in the docs, and I guess there is a values reference for this chart which is very confusing to read https://goteleport.com/docs/kubernetes-access/helm/reference/#teleport-kube-agent but I don't see any specific docs on the agent charts.

> Also, if it's possible to migrate from non-stateful setup, we should document how. If it's not possible, we need a warning somewhere.

It's not really a migration, when not using the volume, every startup of teleport is a new installation. When adding the persistent volume, the first start of the pod will then "install" teleport by persisting the /var/lib/teleport directories, so a subsequent creation of the pod now just has the data from the last startup. So I don't think there is really a migration here other than you no longer need long lived tokens (or like us, have to manually reset the token every time the pod restarts).

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml (hunk 4 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/templates/statefulset.yaml:5]: Added in 1fa70acd5e5bc3ad7a2422a235fe9ccaff8a9657

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml (hunk 3 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/templates/statefulset.yaml]: Address in 1fa70acd5e5bc3ad7a2422a235fe9ccaff8a9657

File: examples/chart/teleport-kube-agent/templates/deployment.yaml (hunk 2 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/templates/deployment.yaml]: Not sure how I missed that one. Address in 1fa70acd5e5bc3ad7a2422a235fe9ccaff8a9657

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml (hunk 4 above)
Comment (webvictim) [reply examples/chart/teleport-kube-agent/templates/statefulset.yaml:5]: We could probably actually achieve this with a Helm template/function - basically define the main chart structure in one place and allow optional includes specific to `Deployment` and `StatefulSet`. That way the code only lives in one file and any changes will be represented in both places.

File: examples/chart/teleport-kube-agent/values.yaml (hunk 1 above)
Comment (webvictim) [reply examples/chart/teleport-kube-agent/values.yaml:69]: I think this volume can be fairly small - no logs are really kept locally (they all go to Docker/containerd) and the sqlite database for the agent itself should only be holding a keypair and some cache data.

**Review by webvictim — COMMENTED** [review]
There aren't decent docs for the `teleport-kube-agent` chart AFAIK - the only real reference is under https://goteleport.com/docs/kubernetes-access/controls/#kubernetes-labels.

@inertial-frame @benarent Do we have plans to write guides for using `teleport-kube-agent` similar to those for `teleport-cluster`?

I think that to merge this, we would at least need to add the additional values to the chart reference - unclear or not, the reference is the canonical source for all the chart's options.

**Comment by awly** [comment]
@knisbet @webvictim yeah, we can just expand comments in `values.yaml` and add a section in https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-kube-agent/README.md

File: examples/chart/teleport-kube-agent/templates/statefulset.yaml (hunk 4 above)
Comment (knisbet) [reply examples/chart/teleport-kube-agent/templates/statefulset.yaml:5]: I believe this is what we were discussing, and this is the first way I wrote it. Unfortunately there are 3, maybe 4, differences that occur so it creates a bit of a messy template that removes/adds sections based on whether it's a StatefulSet or Deployment.

**Comment by knisbet** [comment]
> @knisbet @webvictim yeah, we can just expand comments in `values.yaml` and add a section in https://github.com/gravitational/teleport/blob/master/examples/chart/teleport-kube-agent/README.md

Added some more text in 25533c55cad5bf05a6260baed45b58d59c07c55a

Mainly trying to follow how it's already setup.

**Review by awly — APPROVED** [review]

File: examples/chart/teleport-kube-agent/values.yaml [hunk 5]
```diff
@@ -57,6 +57,26 @@ podSecurityPolicy:
 # Labels is a map of key values pairs about this cluster
 labels: {}
 
+################################################################
+# Values that must be provided if using persistent storage for Teleport state.
+#
+# Assigning a persistent volume to Teleport agent allows the agent to store it's security association with the Teleport
```
Comment (webvictim) [inline examples/chart/teleport-kube-agent/values.yaml:63]:
```suggestion
# Assigning a persistent volume to Teleport agent allows the agent to store its security association with the Teleport
```

File: examples/chart/teleport-kube-agent/values.yaml [hunk 6]
```diff
@@ -57,6 +57,26 @@ podSecurityPolicy:
...
+# cluster for re-use when the pod is restarted. Without a persistent storage for this state, every time Teleport agent
+# starts it must use the authToken to create a new registration with the cluster. By using the persistent volume the
+# authToken can be routinely rotated without breaking agents ability to restart, as the token is only used on the first
```
Comment (webvictim) [inline examples/chart/teleport-kube-agent/values.yaml:66]:
```suggestion
# authToken can be routinely rotated without breaking agents' ability to restart, as the token is only used on the first
```

File: examples/chart/teleport-kube-agent/values.yaml [hunk 7]
```diff
@@ -88,6 +108,10 @@ logLevel: INFO
 # https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity
 affinity: {}
 
+# nodeSelector to apply for pod assignment
+# https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
+nodeSelector: {}
```
Comment (webvictim) [inline examples/chart/teleport-kube-agent/values.yaml:113]: Please add this to the correct section of https://github.com/gravitational/teleport/blob/master/docs/pages/kubernetes-access/helm/reference.mdx as well to keep everything documented. You can just copy a section and change the appropriate parts.

Comment (knisbet) [reply examples/chart/teleport-kube-agent/values.yaml:113]: Alright, I added to the reference doc.

File: docs/pages/kubernetes-access/helm/reference.mdx [hunk 8]
```diff
@@ -1554,6 +1627,32 @@ Kubernetes affinity to set for pod assignments.
...
+`nodeSelector` can be used to add a map of key-value pairs to constrain the nodes the agent pods will run on. 
+
+[Kubernetes reference](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/)
+
+<Tabs>
+  <TabItem label="values.yaml">
+  ```yaml
+  labels:
+    role: node
+    region: us-east
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference.mdx:1645]: Needs to be updated to be `nodeSelector` with the correct syntax.

File: docs/pages/kubernetes-access/helm/reference.mdx [hunk 9]
```diff
@@ -1554,6 +1627,32 @@ Kubernetes affinity to set for pod assignments.
...
+  ```
+  </TabItem>
+  <TabItem label="--set">
+  ```bash
+  --set labels.role=node \
+  --set labels.region=us-east
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference.mdx:1651]: Needs to be updated to be `nodeSelector` with the correct syntax.

File: docs/pages/kubernetes-access/helm/reference.mdx [hunk 10]
```diff
@@ -1326,6 +1326,79 @@ These labels can then be used with Teleport's RBAC policies to define access rul
...
 
+## `storage`
+
+### `storage.enabled`
+
+| Type | Default value |
+| - | - |
+| `bool` | `false` |
+
+Enables the creation of a kubernetes persistent volume to hold teleport agent state.
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference.mdx:1337]:
```suggestion
Enables the creation of a Kubernetes persistent volume to hold Teleport agent state.
```

File: docs/pages/kubernetes-access/helm/reference.mdx [hunk 11]
```diff
@@ -1326,6 +1326,79 @@ These labels can then be used with Teleport's RBAC policies to define access rul
...
+</Tabs>
+
+### `storage.storageClassName`
+
+| Type | Default value |
+| - | - |
+| `string` | `nil` |
+
+The storage class name the persistent volume should use when creating persistent volume claims. The provided storage class
+name needs to exist on the kubernetes cluster for Teleport to use.
```
Comment (webvictim) [inline docs/pages/kubernetes-access/helm/reference.mdx:1362]:
```suggestion
name needs to exist on the Kubernetes cluster for Teleport to use.
```

**Review by webvictim — APPROVED** [review]
LGTM with a few nits

**Review by russjones — APPROVED** [review]
Bot.
